<a href="https://colab.research.google.com/github/Palvamajay/Exploratory-data-analysis/blob/main/EDA_2_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **EDA 2 Assignment**

### **Importing the libraries**

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df=pd.read_csv('/content/adult_with_headers.csv')
df.head()

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital_gain  capital_loss  hours_per_week  native_country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [ ]:
df.shape

(32561, 15)

from the above we have the `32561 rows` and the `15 columns`

In [ ]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'income'],
      dtype='object')

In [ ]:
df.nunique()

age                  73
workclass             9
fnlwgt            21648
education            16
education_num        16
marital_status        7
occupation           15
relationship          6
race                  5
sex                   2
capital_gain        119
capital_loss         92
hours_per_week       94
native_country       42
income                2
dtype: int64

In [ ]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education_num      int64
marital_status    object
occupation        object
relationship      object
race              object
sex               object
capital_gain       int64
capital_loss       int64
hours_per_week     int64
native_country    object
income            object
dtype: object

In [ ]:
df.info() # checking the information about the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [ ]:
df.isnull().sum().sum() # checking for the null values

0

In [ ]:
for col in df.columns:
  if df[col].dtype=="object":
    print("*"*30)
    print(f"the value count of {col}")
    print(df[col].value_counts())
    print("*"*30)

******************************
the value count of workclass
workclass
Private             22696
Self-emp-not-inc     2541
Local-gov            2093
?                    1836
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: count, dtype: int64
******************************
******************************
the value count of education
education
HS-grad         10501
Some-college     7291
Bachelors        5355
Masters          1723
Assoc-voc        1382
11th             1175
Assoc-acdm       1067
10th              933
7th-8th           646
Prof-school       576
9th               514
12th              433
Doctorate         413
5th-6th           333
1st-4th           168
Preschool          51
Name: count, dtype: int64
******************************
******************************
the value count of marital_status
marital_status
Married-civ-spouse       14976
Never-married            10683
Divorced          

In [ ]:
df.duplicated().sum()

24

In [ ]:
df.drop_duplicates(inplace=True)

In [ ]:
df.reset_index(drop=True,inplace=True)

In [ ]:
df.shape

(32537, 15)

From the above given data there is no null values in this there is only the duplicate values in this so we have droped the duplicated values fron this data for the feature process

### **Scaling the data**

In [ ]:
num_col=df.select_dtypes(include=[np.number])

In [ ]:
num_col.head()

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week
0   39   77516             13          2174             0              40
1   50   83311             13             0             0              13
2   38  215646              9             0             0              40
3   53  234721              7             0             0              40
4   28  338409             13             0             0              40

## **Encoding the categorical columns**

### **----> pd.get_dummies**

In [ ]:
data1=df.copy()  # copying the data for the encoding

In [ ]:
data1=pd.get_dummies(data1, dtype=int)
data1

age  fnlwgt  education_num  capital_gain  capital_loss  hours_per_week  \
0       39   77516             13          2174             0              40   
1       50   83311             13             0             0              13   
2       38  215646              9             0             0              40   
3       53  234721              7             0             0              40   
4       28  338409             13             0             0              40   
...    ...     ...            ...           ...           ...             ...   
32532   27  257302             12             0             0              38   
32533   40  154374              9             0             0              40   
32534   58  151910              9             0             0              40   
32535   22  201490              9             0             0              20   
32536   52  287927              9         15024             0              40   

       workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  \
0                 0                       0                     0   
1                 0                       0                     0   
2                 0                       0                     0   
3                 0                       0                     0   
4                 0                       0                     0   
...             ...                     ...                   ...   
32532             0                       0                     0   
32533             0                       0                     0   
32534             0                       0                     0   
32535             0                       0                     0   
32536             0                       0                     0   

       workclass_ Never-worked  ...  native_country_ Scotland  \
0                            0  ...                         0   
1                            0  ...                         0   
2                            0  ...                         0   
3                            0  ...                         0   
4                            0  ...                         0   
...                        ...  ...                       ...   
32532                        0  ...                         0   
32533                        0  ...                         0   
32534                        0  ...                         0   
32535                        0  ...                         0   
32536                        0  ...                         0   

       native_country_ South  native_country_ Taiwan  \
0                          0                       0   
1                          0                       0   
2                          0                       0   
3                          0                       0   
4                          0                       0   
...                      ...                     ...   
32532                      0                       0   
32533                      0                       0   
32534                      0                       0   
32535                      0                       0   
32536                      0                       0   

       native_country_ Thailand  native_country_ Trinadad&Tobago  \
0                             0                                0   
1                             0                                0   
2                             0                                0   
3                             0                                0   
4                             0                                0   
...                         ...                              ...   
32532                         0                                0   
32533                         0                                0   
32534                         0                                0   
32535                         0                                0   
325

From the above we have encoded the variables with the help of the `pd.get_dummies` method

### **----->Onehot_encoding**

In [ ]:
data2=df.copy()
data2

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32532   27            Private  257302   Assoc-acdm             12   
32533   40            Private  154374      HS-grad              9   
32534   58            Private  151910      HS-grad              9   
32535   22            Private  201490      HS-grad              9   
32536   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32532   Married-civ-spouse        Tech-support            Wife   White   
32533   Married-civ-spouse   Machine-op-inspct         Husband   White   
32534              Widowed        Adm-clerical       Unmarried   White   
32535        Never-married        Adm-clerical       Own-child   White   
32536   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32532   Female             0             0              38   United-States   
32533     Male             0             0              40   United-States   
32534   Female             0             0              40   United-States   
32535     Male             0             0              20   United-States   
32536   Female         15024             0              40   United-States   

       income  
0       <=50K  
1       <=50K  
2       <=50K  
3       <=50K  
4       <=50K  
...       ...  
32532   <=50K  
32533    >50K  
32534   <=50K  
32535   <=50K  
32536    >50K  

[32537 rows x 15 columns]

### **We need to import the librari for the onehot encoding**

In [ ]:
from sklearn.preprocessing import OneHotEncoder

In [ ]:
oe=OneHotEncoder()
data2=oe.fit_transform(data2).toarray()
data2

array([[0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       ...,
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [ ]:
data2=pd.DataFrame(data2,columns=oe.get_feature_names_out())
data2

age_17  age_18  age_19  age_20  age_21  age_22  age_23  age_24  age_25  \
0         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
1         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...       ...     ...     ...     ...     ...     ...     ...     ...     ...   
32532     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
32533     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
32534     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
32535     0.0     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   
32536     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

       age_26  ...  native_country_ Scotland  native_country_ South  \
0         0.0  ...                       0.0                    0.0   
1         0.0  ...                       0.0                    0.0   
2         0.0  ...                       0.0                    0.0   
3         0.0  ...                       0.0                    0.0   
4         0.0  ...                       0.0                    0.0   
...       ...  ...                       ...                    ...   
32532     0.0  ...                       0.0                    0.0   
32533     0.0  ...                       0.0                    0.0   
32534     0.0  ...                       0.0                    0.0   
32535     0.0  ...                       0.0                    0.0   
32536     0.0  ...                       0.0                    0.0   

       native_country_ Taiwan  native_country_ Thailand  \
0                         0.0                       0.0   
1                         0.0                       0.0   
2                         0.0                       0.0   
3                         0.0                       0.0   
4                         0.0                       0.0   
...                       ...                       ...   
32532                     0.0                       0.0   
32533                     0.0                       0.0   
32534                     0.0                       0.0   
32535                     0.0                       0.0   
32536                     0.0                       0.0   

       native_country_ Trinadad&Tobago  native_country_ United-States  \
0                                  0.0                            1.0   
1                                  0.0                            1.0   
2                                  0.0                            1.0   
3                                  0.0                            1.0   
4                                  0.0                            0.0   
...                                ...                            ...   
32532                              0.0                            1.0   
32533                              0.0                            1.0   
32534                              0.0                            1.0   
32535                              0.0                            1.0   
32536                              0.0                            1.0   

       native_country_ Vietnam  native_country_ Yugoslavia  income_ <=50K  \
0                          0.0                         0.0            1.0   
1                          0.0                         0.0            1.0   
2                          0.0                         0.0            1.0   
3                          0.0                         0.0            1.0   
4                          0.0                         0.0            1.0   
...                        ...                         ...            ...   
32532                      0.0                         0

**By usin the onehot encoding there is lot of seperation of the columns.it create the many columns acccording the values in the columns that leads to very high dimension data**

### **--->Lable_encoding**

In [ ]:
from sklearn.preprocessing import LabelEncoder

In [ ]:
data3=df.copy()

In [ ]:
le=LabelEncoder()
for col in data3.select_dtypes(include=['object']).columns:
    data3[col] = le.fit_transform(data3[col])

In [ ]:
data3

age  workclass  fnlwgt  education  education_num  marital_status  \
0       39          7   77516          9             13               4   
1       50          6   83311          9             13               2   
2       38          4  215646         11              9               0   
3       53          4  234721          1              7               2   
4       28          4  338409          9             13               2   
...    ...        ...     ...        ...            ...             ...   
32532   27          4  257302          7             12               2   
32533   40          4  154374         11              9               2   
32534   58          4  151910         11              9               6   
32535   22          4  201490         11              9               4   
32536   52          5  287927         11              9               2   

       occupation  relationship  race  sex  capital_gain  capital_loss  \
0               1             1     4    1          2174             0   
1               4             0     4    1             0             0   
2               6             1     4    1             0             0   
3               6             0     2    1             0             0   
4              10             5     2    0             0             0   
...           ...           ...   ...  ...           ...           ...   
32532          13             5     4    0             0             0   
32533           7             0     4    1             0             0   
32534           1             4     4    0             0             0   
32535           1             3     4    1             0             0   
32536           4             5     4    0         15024             0   

       hours_per_week  native_country  income  
0                  40              39       0  
1                  13              39       0  
2                  40              39       0  
3                  40              39       0  
4                  40               5       0  
...               ...             ...     ...  
32532              38              39       0  
32533              40              39       1  
32534              40              39       0  
32535              20              39       0  
32536              40              39       1  

[32537 rows x 15 columns]

In [ ]:
for col in data3.columns:
  if data3[col].dtype=="int":
    print("*"*30)
    print(f"the value count of {col}")
    print(data3[col].value_counts())
    print("*"*30)

******************************
the value count of age
age
36    898
31    888
34    886
23    876
33    875
     ... 
83      6
88      3
85      3
86      1
87      1
Name: count, Length: 73, dtype: int64
******************************
******************************
the value count of workclass
workclass
4    22673
6     2540
2     2093
0     1836
7     1298
5     1116
1      960
8       14
3        7
Name: count, dtype: int64
******************************
******************************
the value count of fnlwgt
fnlwgt
164190    13
123011    13
203488    13
148995    12
113364    12
          ..
138342     1
158040     1
232784     1
325573     1
257302     1
Name: count, Length: 21648, dtype: int64
******************************
******************************
the value count of education
education
11    10494
15     7282
9      5353
12     1722
8      1382
1      1175
7      1067
0       933
5       645
14      576
6       514
2       433
10      413
4       332
3       166
13      

**Both One-Hot Encoding `(OHE)` and Label Encoding `(LE)` are techniques for converting categorical data into numerical form for machine learning models. However, they have different use cases, advantages, and disadvantages.**

**one-hot encoding : Good for: Categories with no natural order (e.g., colors, cities)**  
**Label encoding : Good for: Ordered categories (e.g., Low, Medium, High)**


### **Scaling the data**

### **Standardzation**

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

In [ ]:
scale=StandardScaler()
data3=scale.fit_transform(data3)
data3

array([[ 0.03038995,  2.14997135, -1.06356884, ..., -0.03566374,
         0.29133547, -0.56337742],
       [ 0.8369732 ,  1.46335725, -1.00866848, ..., -2.22248299,
         0.29133547, -0.56337742],
       [-0.0429358 ,  0.09012905,  0.24503992, ..., -0.03566374,
         0.29133547, -0.56337742],
       ...,
       [ 1.42357919,  0.09012905, -0.35877886, ..., -0.03566374,
         0.29133547, -0.56337742],
       [-1.2161478 ,  0.09012905,  0.11092955, ..., -1.65552985,
         0.29133547, -0.56337742],
       [ 0.9836247 ,  0.77674315,  0.92981187, ..., -0.03566374,
         0.29133547,  1.77500901]])

In [ ]:
df1=pd.DataFrame(data3,columns=df.columns)
df1

age  workclass    fnlwgt  education  education_num  \
0      0.030390   2.149971 -1.063569  -0.335266       1.134777   
1      0.836973   1.463357 -1.008668  -0.335266       1.134777   
2     -0.042936   0.090129  0.245040   0.181519      -0.420679   
3      1.056950   0.090129  0.425752  -2.402406      -1.198407   
4     -0.776193   0.090129  1.408066  -0.335266       1.134777   
...         ...        ...       ...        ...            ...   
32532 -0.849519   0.090129  0.639678  -0.852051       0.745913   
32533  0.103716   0.090129 -0.335436   0.181519      -0.420679   
32534  1.423579   0.090129 -0.358779   0.181519      -0.420679   
32535 -1.216148   0.090129  0.110930   0.181519      -0.420679   
32536  0.983625   0.776743  0.929812   0.181519      -0.420679   

       marital_status  occupation  relationship      race       sex  \
0            0.921857   -1.317629     -0.277864  0.393685  0.703020   
1           -0.405919   -0.608318     -0.900126  0.393685  0.703020   
2           -1.733696   -0.135444     -0.277864  0.393685  0.703020   
3           -0.405919   -0.135444     -0.900126 -1.962488  0.703020   
4           -0.405919    0.810304      2.211186 -1.962488 -1.422436   
...               ...         ...           ...       ...       ...   
32532       -0.405919    1.519615      2.211186  0.393685 -1.422436   
32533       -0.405919    0.100993     -0.900126  0.393685  0.703020   
32534        2.249634   -1.317629      1.588923  0.393685 -1.422436   
32535        0.921857   -1.317629      0.966661  0.393685  0.703020   
32536       -0.405919   -0.608318      2.211186  0.393685 -1.422436   

       capital_gain  capital_loss  hours_per_week  native_country    income  
0          0.148292     -0.216743       -0.035664        0.291335 -0.563377  
1         -0.145975     -0.216743       -2.222483        0.291335 -0.563377  
2         -0.145975     -0.216743       -0.035664        0.291335 -0.563377  
3         -0.145975     -0.216743       -0.035664        0.291335 -0.563377  
4         -0.145975     -0.216743       -0.035664       -4.056151 -0.563377  
...             ...           ...             ...             ...       ...  
32532     -0.145975     -0.216743       -0.197650        0.291335 -0.563377  
32533     -0.145975     -0.216743       -0.035664        0.291335  1.775009  
32534     -0.145975     -0.216743       -0.035664        0.291335 -0.563377  
32535     -0.145975     -0.216743       -1.655530        0.291335 -0.563377  
32536      1.887635     -0.216743       -0.035664        0.291335  1.775009  

[32537 rows x 15 columns]

### **Normalazation**

In [ ]:
nz=MinMaxScaler()
data3=nz.fit_transform(data3)
data3

array([[0.30136986, 0.875     , 0.0443019 , ..., 0.39795918, 0.95121951,
        0.        ],
       [0.45205479, 0.75      , 0.0482376 , ..., 0.12244898, 0.95121951,
        0.        ],
       [0.28767123, 0.5       , 0.13811345, ..., 0.39795918, 0.95121951,
        0.        ],
       ...,
       [0.56164384, 0.5       , 0.09482688, ..., 0.39795918, 0.95121951,
        0.        ],
       [0.06849315, 0.5       , 0.12849934, ..., 0.19387755, 0.95121951,
        0.        ],
       [0.47945205, 0.625     , 0.18720338, ..., 0.39795918, 0.95121951,
        1.        ]])

In [ ]:
df2=pd.DataFrame(data3,columns=df.columns)
df2

age  workclass    fnlwgt  education  education_num  \
0      0.301370      0.875  0.044302   0.600000       0.800000   
1      0.452055      0.750  0.048238   0.600000       0.800000   
2      0.287671      0.500  0.138113   0.733333       0.533333   
3      0.493151      0.500  0.151068   0.066667       0.400000   
4      0.150685      0.500  0.221488   0.600000       0.800000   
...         ...        ...       ...        ...            ...   
32532  0.136986      0.500  0.166404   0.466667       0.733333   
32533  0.315068      0.500  0.096500   0.733333       0.533333   
32534  0.561644      0.500  0.094827   0.733333       0.533333   
32535  0.068493      0.500  0.128499   0.733333       0.533333   
32536  0.479452      0.625  0.187203   0.733333       0.533333   

       marital_status  occupation  relationship  race  sex  capital_gain  \
0            0.666667    0.071429           0.2   1.0  1.0      0.021740   
1            0.333333    0.285714           0.0   1.0  1.0      0.000000   
2            0.000000    0.428571           0.2   1.0  1.0      0.000000   
3            0.333333    0.428571           0.0   0.5  1.0      0.000000   
4            0.333333    0.714286           1.0   0.5  0.0      0.000000   
...               ...         ...           ...   ...  ...           ...   
32532        0.333333    0.928571           1.0   1.0  0.0      0.000000   
32533        0.333333    0.500000           0.0   1.0  1.0      0.000000   
32534        1.000000    0.071429           0.8   1.0  0.0      0.000000   
32535        0.666667    0.071429           0.6   1.0  1.0      0.000000   
32536        0.333333    0.285714           1.0   1.0  0.0      0.150242   

       capital_loss  hours_per_week  native_country  income  
0               0.0        0.397959        0.951220     0.0  
1               0.0        0.122449        0.951220     0.0  
2               0.0        0.397959        0.951220     0.0  
3               0.0        0.397959        0.951220     0.0  
4               0.0        0.397959        0.121951     0.0  
...             ...             ...             ...     ...  
32532           0.0        0.377551        0.951220     0.0  
32533           0.0        0.397959        0.951220     1.0  
32534           0.0        0.397959        0.951220     0.0  
32535           0.0        0.193878        0.951220     0.0  
32536           0.0        0.397959        0.951220     1.0  

[32537 rows x 15 columns]

**standarzation**:This is used when the data has the different scales such as the age vs the salary columns  
It follows the normal distribution

Normalazation : It does not follows the normal diistribution the scaled values are in the range between the `0` and `1`

## **Feature Engineering:**

In [ ]:
df['income_per_hour'] = df['capital_gain'] / (df['hours_per_week'] )  # Avoid division by zero


In [ ]:
df

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32532   27            Private  257302   Assoc-acdm             12   
32533   40            Private  154374      HS-grad              9   
32534   58            Private  151910      HS-grad              9   
32535   22            Private  201490      HS-grad              9   
32536   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32532   Married-civ-spouse        Tech-support            Wife   White   
32533   Married-civ-spouse   Machine-op-inspct         Husband   White   
32534              Widowed        Adm-clerical       Unmarried   White   
32535        Never-married        Adm-clerical       Own-child   White   
32536   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32532   Female             0             0              38   United-States   
32533     Male             0             0              40   United-States   
32534   Female             0             0              40   United-States   
32535     Male             0             0              20   United-States   
32536   Female         15024             0              40   United-States   

       income  income_per_hour  
0       <=50K            54.35  
1       <=50K             0.00  
2       <=50K             0.00  
3       <=50K             0.00  
4       <=50K             0.00  
...       ...              ...  
32532   <=50K             0.00  
32533    >50K             0.00  
32534   <=50K             0.00  
32535   <=50K             0.00  
32536    >50K           375.60  

[32537 rows x 16 columns]

In [ ]:
df['age_group'] = pd.cut(df['age'], bins=[0, 18, 35, 50, 65, 100], labels=['Teen', 'Young', 'Adult', 'Middle_Age', 'Senior'])


In [ ]:
df

age          workclass  fnlwgt    education  education_num  \
0       39          State-gov   77516    Bachelors             13   
1       50   Self-emp-not-inc   83311    Bachelors             13   
2       38            Private  215646      HS-grad              9   
3       53            Private  234721         11th              7   
4       28            Private  338409    Bachelors             13   
...    ...                ...     ...          ...            ...   
32532   27            Private  257302   Assoc-acdm             12   
32533   40            Private  154374      HS-grad              9   
32534   58            Private  151910      HS-grad              9   
32535   22            Private  201490      HS-grad              9   
32536   52       Self-emp-inc  287927      HS-grad              9   

            marital_status          occupation    relationship    race  \
0            Never-married        Adm-clerical   Not-in-family   White   
1       Married-civ-spouse     Exec-managerial         Husband   White   
2                 Divorced   Handlers-cleaners   Not-in-family   White   
3       Married-civ-spouse   Handlers-cleaners         Husband   Black   
4       Married-civ-spouse      Prof-specialty            Wife   Black   
...                    ...                 ...             ...     ...   
32532   Married-civ-spouse        Tech-support            Wife   White   
32533   Married-civ-spouse   Machine-op-inspct         Husband   White   
32534              Widowed        Adm-clerical       Unmarried   White   
32535        Never-married        Adm-clerical       Own-child   White   
32536   Married-civ-spouse     Exec-managerial            Wife   White   

           sex  capital_gain  capital_loss  hours_per_week  native_country  \
0         Male          2174             0              40   United-States   
1         Male             0             0              13   United-States   
2         Male             0             0              40   United-States   
3         Male             0             0              40   United-States   
4       Female             0             0              40            Cuba   
...        ...           ...           ...             ...             ...   
32532   Female             0             0              38   United-States   
32533     Male             0             0              40   United-States   
32534   Female             0             0              40   United-States   
32535     Male             0             0              20   United-States   
32536   Female         15024             0              40   United-States   

       income  income_per_hour   age_group  
0       <=50K            54.35       Adult  
1       <=50K             0.00       Adult  
2       <=50K             0.00       Adult  
3       <=50K             0.00  Middle_Age  
4       <=50K             0.00       Young  
...       ...              ...         ...  
32532   <=50K             0.00       Young  
32533    >50K             0.00       Adult  
32534   <=50K             0.00  Middle_Age  
32535   <=50K             0.00       Young  
32536    >50K           375.60  Middle_Age  

[32537 rows x 17 columns]

With the help of the above data we have created the two new columns they are the `income_per_hour`	`age_group`
`income_per_hour`  the  reason for creating this column is that the we can check the average income of a person  
`age_group` this is to describe about the age in the given columns in the form of the `'Teen', 'Young', 'Adult', 'Middle_Age', 'Senior'`

**Appilying the log transformaion to the captial_gain**

skewness befor transformation

In [ ]:
print("Skewness before:", df['capital_gain'].skew())


Skewness before: -0.08404028007521744


In [ ]:
df['capital_gain']=np.log(df['capital_gain'])

skewness after transformation

In [ ]:
print("Skewness after:", df['capital_gain'].skew())

Skewness after: -0.47903618620528277


"I applied a log transformation to the `capital_gain` feature because it was highly right-skewed, meaning most values were small while a few were extremely large. This transformation helped normalize the distribution, making it more suitable for models that assume normality, such as linear regression."

## **Feature Selection:**

In [ ]:
from sklearn.ensemble import IsolationForest

In [ ]:
iso_forest=IsolationForest(contamination=0.05,random_state=42)
iso_forest.fit(data3)

IsolationForest(contamination=0.05, random_state=42)

In [ ]:
pred=iso_forest.predict(data3)
pred

array([ 1,  1,  1, ...,  1,  1, -1])

In [ ]:
data3['outliers']=pred
data3

age  workclass    fnlwgt  education  education_num  \
0      0.301370      0.875  0.044302   0.600000       0.800000   
1      0.452055      0.750  0.048238   0.600000       0.800000   
2      0.287671      0.500  0.138113   0.733333       0.533333   
3      0.493151      0.500  0.151068   0.066667       0.400000   
4      0.150685      0.500  0.221488   0.600000       0.800000   
...         ...        ...       ...        ...            ...   
32532  0.136986      0.500  0.166404   0.466667       0.733333   
32533  0.315068      0.500  0.096500   0.733333       0.533333   
32534  0.561644      0.500  0.094827   0.733333       0.533333   
32535  0.068493      0.500  0.128499   0.733333       0.533333   
32536  0.479452      0.625  0.187203   0.733333       0.533333   

       marital_status  occupation  relationship  race  sex  capital_gain  \
0            0.666667    0.071429           0.2   1.0  1.0      0.021740   
1            0.333333    0.285714           0.0   1.0  1.0      0.000000   
2            0.000000    0.428571           0.2   1.0  1.0      0.000000   
3            0.333333    0.428571           0.0   0.5  1.0      0.000000   
4            0.333333    0.714286           1.0   0.5  0.0      0.000000   
...               ...         ...           ...   ...  ...           ...   
32532        0.333333    0.928571           1.0   1.0  0.0      0.000000   
32533        0.333333    0.500000           0.0   1.0  1.0      0.000000   
32534        1.000000    0.071429           0.8   1.0  0.0      0.000000   
32535        0.666667    0.071429           0.6   1.0  1.0      0.000000   
32536        0.333333    0.285714           1.0   1.0  0.0      0.150242   

       capital_loss  hours_per_week  native_country  income  income_per_hour  \
0               0.0        0.397959        0.951220     0.0              1.0   
1               0.0        0.122449        0.951220     0.0              1.0   
2               0.0        0.397959        0.951220     0.0              1.0   
3               0.0        0.397959        0.951220     0.0              1.0   
4               0.0        0.397959        0.121951     0.0             -1.0   
...             ...             ...             ...     ...              ...   
32532           0.0        0.377551        0.951220     0.0              1.0   
32533           0.0        0.397959        0.951220     1.0              1.0   
32534           0.0        0.397959        0.951220     0.0              1.0   
32535           0.0        0.193878        0.951220     0.0              1.0   
32536           0.0        0.397959        0.951220     1.0             -1.0   

       age_group  outliers  
0            1.0         1  
1            1.0         1  
2            1.0         1  
3            1.0         1  
4           -1.0        -1  
...          ...       ...  
32532        1.0         1  
32533        1.0         1  
32534        1.0         1  
32535        1.0         1  
32536       -1.0        -1  

[32537 rows x 18 columns]

In [ ]:
data3[data3['outliers']== -1]

age  workclass    fnlwgt  education  education_num  \
4      0.150685      0.500  0.221488   0.600000       0.800000   
11     0.178082      0.875  0.087619   0.600000       0.800000   
14     0.315068      0.500  0.074359   0.533333       0.666667   
15     0.232877      0.500  0.158380   0.333333       0.200000   
27     0.506849      0.000  0.114048   1.000000       0.600000   
...         ...        ...       ...        ...            ...   
32507  0.178082      0.000  0.014619   0.600000       0.800000   
32509  0.506849      0.500  0.221205   0.600000       0.800000   
32514  0.287671      0.500  0.086181   0.600000       0.800000   
32515  0.739726      0.000  0.186826   0.666667       1.000000   
32536  0.479452      0.625  0.187203   0.733333       0.533333   

       marital_status  occupation  relationship  race  sex  capital_gain  \
4            0.333333    0.714286           1.0  0.50  0.0      0.000000   
11           0.333333    0.714286           0.0  0.25  1.0      0.000000   
14           0.333333    0.214286           0.0  0.25  1.0      0.000000   
15           0.333333    1.000000           0.0  0.00  1.0      0.000000   
27           0.333333    0.000000           0.0  0.25  1.0      0.000000   
...               ...         ...           ...   ...  ...           ...   
32507        0.666667    0.000000           0.2  0.25  0.0      0.000000   
32509        0.333333    0.285714           0.0  0.25  1.0      0.000000   
32514        0.000000    0.714286           0.8  0.50  0.0      0.150202   
32515        0.333333    0.000000           0.0  1.00  1.0      0.000000   
32536        0.333333    0.285714           1.0  1.00  0.0      0.150242   

       capital_loss  hours_per_week  native_country  income  income_per_hour  \
4               0.0        0.397959        0.121951     0.0             -1.0   
11              0.0        0.397959        0.463415     1.0             -1.0   
14              0.0        0.397959        0.000000     1.0             -1.0   
15              0.0        0.448980        0.634146     0.0             -1.0   
27              0.0        0.602041        0.853659     1.0             -1.0   
...             ...             ...             ...     ...              ...   
32507           0.0        1.000000        0.951220     0.0             -1.0   
32509           0.0        0.500000        0.585366     1.0             -1.0   
32514           0.0        0.448980        0.951220     1.0             -1.0   
32515           0.0        0.091837        0.951220     1.0             -1.0   
32536           0.0        0.397959        0.951220     1.0             -1.0   

       age_group  outliers  
4           -1.0        -1  
11          -1.0        -1  
14          -1.0        -1  
15          -1.0        -1  
27          -1.0        -1  
...          ...       ...  
32507       -1.0        -1  
32509       -1.0        -1  
32514       -1.0        -1  
32515       -1.0        -1  
32536       -1.0        -1  

[1627 rows x 18 columns]

## **predctive power score (pps)**

In [ ]:
import ppscore as pps

In [ ]:
pps.score(data3, "capital_gain", "education")

{'x': 'capital_gain',
 'y': 'education',
 'ppscore': 0,
 'case': 'regression',
 'is_valid_score': True,
 'metric': 'mean absolute error',
 'baseline_score': 0.18130666666666667,
 'model_score': 0.1940743015975716,
 'model': DecisionTreeRegressor()}